File: 프로그래머스/일반/21년12월시작/디스크 컨트롤러.js
function solution(jobs) {
    var answer = 0;
    let s = -1, n = 0, i = 0;
    const q = [];

    while (i < jobs.length) {
        jobs.map((j) => {
            if (s < j[0] && j[0] <= n) {
                q.push([j[1], j[0]]);
            }
        })

        if (q.length > 0) {
            let x = q.sort(([a, b], [c, d]) => a - c).shift();
            s = n;
            n += x[0];
            answer += n - x[1];
            i++;
        }
        else {
            n++;
        }
    }


    return parseInt(answer / jobs.length);
}

File: 프로그래머스/일반/21년12월시작/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0, s = -1, n = 0, i = 0;
        PriorityQueue<int[]> q = new PriorityQueue<int[]> (new Comparator<int[]> () {
            public int compare(int[] a, int[] b) {
                if (a[1] < b[1])
                    return -1;
                return 1;
            }
        });

        while (i < jobs.length) {
            for (int[] j : jobs) {
                if (s < j[0] && j[0] <= n) {
                    q.add(j);
                }
            }

            if (q.size() > 0){
                int[] x = q.poll();
                s = n;
                n += x[1];
                answer += n - x[0];
                i++;
            }
            else {
                n++;
            }
        }



        return Math.floorDiv(answer, jobs.length);
    }
}

